src/xapp/apphub.8
.TH apphub 8
'''
.SH NAME
\fBapphub\fR \- generic application supervisor
'''
.SH DESCRIPTION
This service spawns applications on user's request. It is meant primarily
for GUI applications, but can handle pretty much anything that does not
require user input from stdin or open TTY on stdout.
.P
\fBapphub\fR is a regular service that should be started during system boot.
'''
.SH USAGE
Applications to be spawned by \fBapphub\fR are defined by executable files
in /etc/apphub/. The base name of the file is what needs to be passed to
\fBappctl\fR in order to spawn the script. When commanded to do so,
\fBapphub\fR execs the script in a newly spawned process.
.P
Sample script, /etc/apphub/foo:
.P
.ni
    #!/sbin/msh

    setenv PATH /usr/bin
    setuid foo:apps

    invoke /usr/bin/xbigapp
.fi
.P
Application defined by this script can them be started using
.P
.ni
    appctl start foo
.fi
.P
Spawned applications get their arguments and environment from \fBappctl\fR.
.P
The scripts inherit credentials from the parent \fBapphub\fR process, which
typically means root uid and full capabilities. It is up to the script to limit
capabilities, set uid and groups, and otherwise set up process environment.
.P
Applications are expected to not take any input from stdin, and to not output
anything to either stdout or stderr during normal operations. In the spawned
process, all three standard file descriptors (0, 1, 2) point to the write end
of a pipe; \fBapphub\fR reads the other end of the pipe into a ring buffer,
and lets the user inspect its contents. The collected output is only meant to
be used for troubleshooting misbehaving applications, it is not supposed to be
a part of normal operations.
.P
The name of the script may be chosen arbitrarily. It is perfectly fine to have
several scripts exec the same executable, possibly with different options or
environment settings.
.P
Applications may exit at will, and generally are expected to do so.
'''
.SH FILES
.IP "/run/ctrl/apphub" 4
Control socket.
.IP "/etc/apphub" 4
Configuration directory.
'''
.SH SEE ALSO
\fBappctl\fR(1).


src/xapp/apphub.c
#include <bits/socket/unix.h>

#include <sys/file.h>
#include <sys/fpath.h>
#include <sys/ppoll.h>
#include <sys/signal.h>
#include <sys/socket.h>
#include <sys/timer.h>
#include <sys/signalfd.h>
#include <sys/prctl.h>

#include <string.h>
#include <sigset.h>
#include <printf.h>
#include <main.h>
#include <util.h>

#include "common.h"
#include "apphub.h"

ERRTAG("apphub");

static void clear_ctrl(void)
{
	sys_unlink(CONTROL);
}

static void kill_all_procs(CTX)
{
	struct proc* pc = ctx->procs;
	struct proc* pe = pc + ctx->nprocs;
	int pid;

	for(; pc < pe; pc++) {
		if((pid = pc->pid) <= 0)
			continue;

		(void)sys_kill(pid, SIGTERM);
	}
}

static void kill_all_conns(CTX)
{
	struct conn* cn = ctx->conns;
	struct conn* ce = cn + ctx->nconns;
	int fd;

	for(; cn < ce; cn++) {
		if((fd = cn->fd) <= 0)
			continue;

		sys_close(fd);
		cn->fd = -1;
	}

	sys_close(ctx->ctlfd);
	ctx->ctlfd = -1;

	ctx->pollset = 0;
}

void quit(const char* msg, char* arg, int err)
{
	clear_ctrl();

	fail(msg, arg, err);
}

static void sigprocmask(int how, sigset_t* mask)
{
	int ret;

	if((ret = sys_sigprocmask(how, mask, NULL)) < 0)
		fail("sigprocmask", NULL, ret);
}

static void open_signalfd(CTX)
{
	int fd;
	sigset_t ss;

	sigemptyset(&ss);

	sigaddset(&ss, SIGINT);
	sigaddset(&ss, SIGHUP);
	sigaddset(&ss, SIGTERM);
	sigaddset(&ss, SIGCHLD);

	sigprocmask(SIG_BLOCK, &ss);

	if((fd = sys_signalfd(-1, &ss, SFD_CLOEXEC)) < 0)
		fail("signalfd", NULL, fd);

	ctx->sigfd = fd;
}

static void setup_control(CTX)
{
	int fd, ret;
	int flags = SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC;
	struct sockaddr_un addr = {
		.family = AF_UNIX,
		.path = CONTROL
	};

	if((fd = sys_socket(AF_UNIX, flags, 0)) < 0)
		fail("socket", "AF_UNIX", fd);

	if((ret = sys_bind(fd, &addr, sizeof(addr))) < 0)
		fail("bind", addr.path, ret);
	else if((ret = sys_listen(fd, 1)))
		quit("listen", addr.path, ret);

	ctx->ctlfd = fd;
}

static void clear_exit(CTX)
{
	clear_ctrl();
	_exit(0xFF);
}

static void maybe_normal_exit(CTX)
{
	if(ctx->timer != TM_STOP)
		return;
	if(ctx->nprocs_running)
		return;

	clear_exit(ctx);
}

static void request_stop(CTX, const char* signame)
{
	if(ctx->timer == TM_STOP) {
		warn(signame, NULL, 0);
		clear_exit(ctx);
	}

	kill_all_procs(ctx);
	kill_all_conns(ctx);

	ctx->timer = TM_STOP;
	ctx->ts.sec = 5;
	ctx->ts.nsec = 0;

	maybe_normal_exit(ctx);
}

static void check_signal(CTX)
{
	struct siginfo si;
	int fd = ctx->sigfd;
	int ret;

	memzero(&si, sizeof(si));

	if((ret = sys_read(fd, &si, sizeof(si))) < 0)
		quit("read", "sigfd", ret);
	if(ret == 0)
		quit("signalfd EOF", NULL, 0);

	int sig = si.signo;

	if(sig == SIGCHLD) {
		check_children(ctx);
		maybe_normal_exit(ctx);
	} else if(sig == SIGINT) {
		request_stop(ctx, "SIGINT");
	} else if(sig == SIGTERM) {
		request_stop(ctx, "SIGTERM");
	} else if(sig == SIGHUP) {
		request_stop(ctx, "SIGHUP");
	}
}

static void handle_timeout(CTX)
{
	int timer = ctx->timer;

	ctx->timer = TM_NONE;

	if(timer == TM_STOP)
		clear_exit(ctx);
	if(timer == TM_MMAP)
		maybe_drop_iobuf(ctx);
}

static void reset_pollfds(CTX)
{
	int npfds = 2 + ctx->nconns_active + ctx->nprocs_nonempty;
	struct pollfd* pfds;
	int need = npfds*sizeof(*pfds);
	void* buf;
	int len;

	ctx->npfds = 0;
	ctx->ptr = ctx->sep;

	if(need < sizeof(ctx->pollbuf))
		goto pbuf;

	if(!(buf = heap_alloc(ctx, need)))
		goto pbuf;

	len = need;
	goto done;
pbuf:
	buf = ctx->pollbuf;
	len = sizeof(ctx->pollbuf);
done:
	ctx->pfds = buf;
	ctx->pfde = buf + len;
}

static void add_poll_fd(CTX, int fd)
{
	int n = ctx->npfds;
	struct pollfd* pfd = &ctx->pfds[n];
	struct pollfd* pfe = ctx->pfde;

	if(pfd + 1 > pfe) {
		ctx->pollset = 0;
		warn("pfds overflow", NULL, 0);
		return;
	}

	pfd->fd = fd;
	pfd->events = POLLIN;

	ctx->npfds = n + 1;
}

static void add_conn_fds(CTX)
{
	struct conn* cn = ctx->conns;
	struct conn* ce = cn + ctx->nconns;
	int fd;

	for(; cn < ce; cn++)
		if((fd = cn->fd) > 0)
			add_poll_fd(ctx, fd);

	ctx->npsep = ctx->npfds;
}

static void add_proc_fds(CTX)
{
	struct proc* pc = ctx->procs;
	struct proc* pe = pc + ctx->nprocs;
	int fd;

	for(; pc < pe; pc++)
		if(!pc->xid)
			continue;
		else if((fd = pc->fd) > 0)
			add_poll_fd(ctx, fd);
}

static void update_poll_fds(CTX)
{
	reset_pollfds(ctx);

	ctx->pollset = 1;

	add_poll_fd(ctx, ctx->ctlfd);
	add_poll_fd(ctx, ctx->sigfd);

	add_conn_fds(ctx);
	add_proc_fds(ctx);

	maybe_trim_heap(ctx);
}

static void process_conns(CTX)
{
	struct pollfd* pfds = ctx->pfds;
	int i = 2, n = ctx->npsep;
	struct conn* cn = ctx->conns;
	struct conn* ce = cn + ctx->nconns;

	for(; i < n; i++) {
		struct pollfd* pf = &pfds[i];
		int revents = pf->revents;
		int polledfd = pf->fd;

		if(!revents)
			continue;

		for(; cn < ce; cn++)
			if(cn->fd == polledfd)
				break;
		if(cn >= ce)
			break;

		if(revents & POLLIN)
			handle_conn(ctx, cn);
		if(revents & ~POLLIN)
			close_conn(ctx, cn);

	} if(i < n) {
		ctx->pollset = 0;
	}
}

static void process_procs(CTX)
{
	struct pollfd* pfds = ctx->pfds;
	int i = ctx->npsep, n = ctx->npfds;
	struct proc* pc = ctx->procs;
	struct proc* pe = pc + ctx->nprocs;

	if(!pc) return;

	for(; i < n; i++) {
		struct pollfd* pf = &pfds[i];
		int revents = pf->revents;
		int polledfd = pf->fd;

		if(!revents)
			continue;

		for(; pc < pe; pc++)
			if(pc->fd == polledfd)
				break;
		if(pc >= pe)
			break;

		if(revents & POLLIN)
			handle_pipe(ctx, pc);
		if(revents & ~POLLIN)
			close_pipe(ctx, pc);

	} if(i < n) {
		ctx->pollset = 0;
	}
}

static void process_events(CTX)
{
	struct pollfd* pfds = ctx->pfds;

	if(pfds[0].revents & POLLIN)
		check_socket(ctx);
	if(pfds[1].revents & POLLIN)
		check_signal(ctx);

	process_conns(ctx);
	process_procs(ctx);
}

static void poll(CTX)
{
	int ret;

	if(!ctx->pollset)
		update_poll_fds(ctx);

	struct pollfd* pfds = ctx->pfds;
	int npfds = ctx->npfds;
	struct timespec* ts = &ctx->ts;

	if(ctx->timer == TM_NONE) ts = NULL;

	if((ret = sys_ppoll(pfds, npfds, ts, NULL)) < 0)
		quit("ppoll", NULL, ret);

	if(ret)
		process_events(ctx);
	else
		handle_timeout(ctx);
}

static void set_subreaper(void)
{
	int ret;

	if((ret = sys_prctl(PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0)) < 0)
		fail("prctl", "PR_SET_CHILD_SUBREAPER", ret);
}

int main(int argc, char** argv)
{
	(void)argv;
	struct top context, *ctx = &context;

	if(argc > 1)
		fail("too many arguments", NULL, 0);

	memzero(ctx, sizeof(*ctx));

	set_subreaper();

	open_signalfd(ctx);
	setup_control(ctx);

	while(1) poll(ctx);
}
